### ISSUE TITLE
Add Dispatch / `libdispatch`

### ISSUE THREAD

**Issue by madsmtm** [madsmtm/objc2#77]
Reasoning:
- Better main thread support.
- `objc2` encoding support (allowing use in framework crates).
- [`async` support](https://github.com/madsmtm/objc2/issues/279)?
- Optimizations (such as the null-ptr optimization, and others).

Desired name: `dispatch2`, since it fits fairly well with the naming scheme (`objc2`, `block2` and `dispatch2` would all be successors to `objc`, `block` and `dispatch`, and would all be libraries internally, as compared to the rest of the crates we expose, which are all frameworks). I've asked for the name [here](https://github.com/marysaka/dispatch2/issues/2), should remember to not close this issue until that repo is archived too.

Unsure how auto-generated we can feasibly be here? But the library is comparatively small compared to most frameworks, so it's probably possible to generate an initial starting point, and then refine later on.

Upstream `dispatch` issues:
- [ ] https://github.com/SSheldon/rust-dispatch/issues/4
- [x] https://github.com/SSheldon/rust-dispatch/issues/7
- [ ] https://github.com/SSheldon/rust-dispatch/issues/14
- [x] https://github.com/SSheldon/rust-dispatch/issues/15
- [ ] https://github.com/SSheldon/rust-dispatch/issues/20

Upstream `dispatch` pull requests to integrate:
- [ ] https://github.com/SSheldon/rust-dispatch/pull/6
- [ ] https://github.com/SSheldon/rust-dispatch/pull/11
- [ ] https://github.com/SSheldon/rust-dispatch/pull/17
- [x] https://github.com/SSheldon/rust-dispatch/pull/21
- [ ] https://github.com/SSheldon/rust-dispatch/pull/22

Integrate ideas from:
- [ ] [`dispatch`](https://github.com/SSheldon/rust-dispatch)
- [ ] [`fruity`'s `dispatch`](https://github.com/nvzqz/fruity/tree/main/src/dispatch)
- [ ] [`dispatchr`](https://github.com/drewcrawford/dispatchr)
- [x] [`dispatch2`](https://github.com/marysaka/dispatch2)
- [ ] [cidre's `dispatch`](https://github.com/yury/cidre/tree/8227b566dfe08063a58dd3e4f557e6582cee7770/cidre/src/dispatch)

Other TODOs:
- [x] Use in framework crates.
- [ ] Refactor such that we have less ugly types lying around.

**Comment by silvanshade** [comment]
I think this would be a good idea. Note that the types are needed for a full translation of the `BackgroundTasks` framework just added. I also find that I am using it quote a lot for working with the `WebKit` framework through `icrate`, though it's conceivable that many of those instances would go away if #279 happens.

**Comment by madsmtm** [comment]
I have integrated [the `dispatch2` crate](https://github.com/marysaka/dispatch2) in https://github.com/madsmtm/objc2/commit/84730a4c6ca77e95e0af4eccc28b2ca4626da5ca. This issue now tracks the upstream `dispatch` issues, as well as making sure we grab all relevant functionality from the other GCD crates.

**Comment by pronebird** [comment]
@madsmtm amazing thank you!

**Comment by yujonglee** [comment]
Hi @madsmtm, does resolving this issue imply that we will have `dispatch_queue_t` support in `objc2`?

**Comment by madsmtm** [comment]
Yeah, I'm working on figuring out hooking it up to `header-translator`, but it requires some changes in `dispatch2`, including making the types ABI-compatible.

Which is currently hard to do, since we have the extra `is_activated` state. [The header says](https://github.com/swiftlang/swift-corelibs-libdispatch/blob/b2a1fad6b49a9394cf864b87ab0034ea8300e2de/dispatch/queue.h#L1094-L1095) that it's undefined behaviour to call `dispatch_set_target_queue` after `dispatch_activate`, so we currently _must_ keep it for soundness, but I suspect the claim to be incorrect, that it isn't language-level UB, just a programmer error.

So I'll investigate further, and try to upstream a documentation update, otherwise we might have to mark some of these functions as `unsafe`.

**Comment by madsmtm** [comment]
Update: I've vastly refactored `dispatch2` in #727, implemented automatic memory management in a6b29cd92acabb14e455c1b4669be394cd1c677b, and managed to get `dispatch2` support in all the framework crates in 91a74f332f3db66378f457ad954f26c2d72277ca!

There's still a bunch of work to be done to make `dispatch2` really nice to use, which this issue still tracks, but that's a large milestone! 🎉

**Comment by madsmtm** [comment]
I improved the bindings a bit more in 3b0377321093d50dd75e5350196e3fd6380edb0a.
